pkg/kubelet/dockertools/docker_manager.go

```diff
@@ -716,13 +720,23 @@ func (dm *DockerManager) runContainer(
 
 // setInfraContainerNetworkConfig sets the network configuration for the infra-container. We only set network configuration for infra-container, all
 // the user containers will share the same network namespace with infra-container.
-func setInfraContainerNetworkConfig(pod *api.Pod, netMode string, opts *kubecontainer.RunContainerOptions, dockerOpts *dockertypes.ContainerCreateConfig) {
+func (dm *DockerManager) setInfraContainerNetworkConfig(pod *api.Pod, netMode string, opts *kubecontainer.RunContainerOptions, dockerOpts *dockertypes.ContainerCreateConfig) {
 	exposedPorts, portBindings := makePortsAndBindings(opts.PortMappings)
 	dockerOpts.Config.ExposedPorts = exposedPorts
 	dockerOpts.HostConfig.PortBindings = dockernat.PortMap(portBindings)
 
+	result, err := dm.checkDockerAPIVersion(dockerV112APIVersion)
+	if err != nil {
+		glog.Errorf("Failed to check docker api version: %v", err)
+	}
+
+	hostNetPlusDNSEnabled := result >= 0
+
 	if netMode != namespaceModeHost {
 		dockerOpts.Config.Hostname = opts.Hostname
+	}
+
+	if netMode != namespaceModeHost || (netMode == namespaceModeHost && hostNetPlusDNSEnabled) {
```

Comment:
> So, if I understand points 3 and 4 correctly, the DOCKER_OPTS were already there but this adds them via the Docker CLI/API ?
Yes, that's right (see https://github.com/docker/docker/issues/21976#issue-147926813).

> To understand the urgency, is there something that is going to break in existing code after the 1.12 Docker release on 28th July ?

There were long discussions and numerious requests in both docker and k8s projects for adding the ability to set dns settings along with --net=host. And now it is going to be available in docker 1.12. But this desirable feature still wouldn't become available to k8s users without additional patch. So it's just about to have it in place asap. :)